ext/oj/dump_compat.c

```diff
@@ -138,6 +138,39 @@ dump_to_json(VALUE obj, Out out) {
     *out->cur = '\0';
 }
 
+static void dump_as_json(VALUE obj, int depth, Out out, bool as_ok) {
+    volatile VALUE ja;
+
+    if (Yes == out->opts->trace) {
+        oj_trace("as_json", obj, __FILE__, __LINE__, depth + 1, TraceRubyIn);
+    }
+    // Some classes elect to not take an options argument so check the arity
+    // of as_json.
+    if (0 == rb_obj_method_arity(obj, oj_as_json_id)) {
+        ja = rb_funcall(obj, oj_as_json_id, 0);
+    } else {
+        ja = rb_funcall2(obj, oj_as_json_id, out->argc, out->argv);
+    }
+    if (Yes == out->opts->trace) {
+        oj_trace("as_json", obj, __FILE__, __LINE__, depth + 1, TraceRubyOut);
+    }
+
+    out->argc = 0;
+    if (ja == obj || !as_ok) {
+        // Once as_json is called it should never be called again on the same
+        // object with as_ok.
+        oj_dump_compat_val(ja, depth, out, false);
+    } else {
+        int type = rb_type(ja);
+
+        if (T_HASH == type || T_ARRAY == type) {
```

Comment:
This does seem to do anything as the exact call and argument are used in both cases.